### ISSUE TITLE
HTTP listener stops working once I add HTTPS listener

### ISSUE THREAD

**Issue by AurimasNav** [cilium/cilium#36750]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.16.4 and lower than v1.17.0

### What happened?

It seems that once I add HTTPS listener the HTTPRoute depending on HTTP listener stops working. This prevents cert-manager issuing certificates for https using http01 solver (and dns solver does not work with our .internal domain).

### How can we reproduce the issue?

1. cilium helm install values:

```yaml
k8sServiceHost: 10.10.0.7
k8sServicePort: 6443
kubeProxyReplacement: true
ipam:
  operator:
    clusterPoolIPv4PodCIDRList: ["10.42.0.0/16"]
operator:
  replicas: 1
gatewayAPI:
  enabled: true
l2announcements:
  enabled: true
```

2. Create gateway with http listener

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: internal
  namespace: cilium
spec:
  infrastructure:
    annotations:
      io.cilium/lb-ipam-ips: 10.10.0.33
  gatewayClassName: cilium
  listeners:
  - name: http-solver
    protocol: HTTP
    port: 80
    allowedRoutes:
        namespaces:
          from: All
```

3. create HTTPRoute

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: hello-world-http
spec:
  hostnames:
  - hello-world.onprem.internal
  parentRefs:
  - name: internal
    namespace: cilium
    sectionName: http-solver
  rules:
  - backendRefs:
    - name: hello-world
      port: 80
``` 

4. verify that httproute is working:

```shell
curl -I http://hello-world.onprem.internal

HTTP/1.1 200 OK
accept-ranges: bytes
content-length: 1395
content-type: text/html; charset=utf-8
last-modified: Tue, 22 Jun 2021 05:40:33 GMT
date: Fri, 20 Dec 2024 17:21:17 GMT
x-envoy-upstream-service-time: 0
server: envoy

```

describe httproute

```yaml
Name:         hello-world-http
Namespace:    hello-world
Labels:       app=hello-world
              environment=dev
              tier=backend
Annotations:  argocd.argoproj.io/tracking-id: hello-world:gateway.networking.k8s.io/HTTPRoute:hello-world/hello-world-http
API Version:  gateway.networking.k8s.io/v1
Kind:         HTTPRoute
Metadata:
  Creation Timestamp:  2024-12-20T16:39:38Z
  Generation:          1
  Resource Version:    3864805
  UID:                 2a7b4144-c397-42ad-a47b-88859e99b309
Spec:
  Hostnames:
    hello-world.onprem.internal
  Parent Refs:
    Group:         gateway.networking.k8s.io
    Kind:          Gateway
    Name:          internal
    Namespace:     cilium
    Section Name:  http-solver
  Rules:
    Backend Refs:
      Group:   
      Kind:    Service
      Name:    hello-world
      Port:    80
      Weight:  1
    Matches:
      Path:
        Type:   PathPrefix
        Value:  /
Status:
  Parents:
    Conditions:
      Last Transition Time:  2024-12-20T16:39:38Z
      Message:               Accepted HTTPRoute
      Observed Generation:   1
      Reason:                Accepted
      Status:                True
      Type:                  Accepted
      Last Transition Time:  2024-12-20T16:39:38Z
      Message:               Service reference is valid
      Observed Generation:   1
      Reason:                ResolvedRefs
      Status:                True
      Type:                  ResolvedRefs
    Controller Name:         io.cilium/gateway-controller
    Parent Ref:
      Group:         gateway.networking.k8s.io
      Kind:          Gateway
      Name:          internal
      Namespace:     cilium
      Section Name:  http-solver
Events:              <none>
```

5. add HTTPS listener to the gateway

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: internal
  namespace: cilium
spec:
  infrastructure:
    annotations:
      io.cilium/lb-ipam-ips: 10.10.0.33
  gatewayClassName: cilium
  listeners:
  - name: http-solver
    protocol: HTTP
    port: 80
    allowedRoutes:
        namespaces:
          from: All
  - name: hello-world
    protocol: HTTPS
    port: 443
    hostname: "hello-world.onprem.internal"
    tls:
      certificateRefs:
      - group: ''
        kind: Secret
        name: hello-world-internal
    allowedRoutes:
      namespaces:
        from: All
```

6. check if service on http is still working as expected

```shell
curl -I http://hello-world.onprem.internal 
                                   
HTTP/1.1 404 Not Found
date: Fri, 20 Dec 2024 17:27:39 GMT
server: envoy
transfer-encoding: chunked
```

### Cilium Version

cilium-cli: v0.16.21 compiled with go1.23.4 on darwin/arm64
cilium image (default): v1.16.4
cilium image (stable): v1.16.5
cilium image (running): unknown. Unable to obtain cilium version. Reason: release: not found
(1.16.5 helm chart installed via argocd)

### Kernel Version

Linux seku-k8s-01 6.8.0-49-generic #49-Ubuntu SMP PREEMPT_DYNAMIC Mon Nov  4 02:06:24 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.31.2
Kustomize Version: v5.4.2
Server Version: v1.30.6+k3s1

### Relevant log output

```shell
cilium envoy pod log


[2024-12-20 17:29:30.048][76][trace][filter] [external/envoy/source/extensions/filters/listener/tls_inspector/tls_inspector.cc:106] tls inspector: new connection accepted
[2024-12-20 17:29:30.048][76][trace][misc] [external/envoy/source/common/network/tcp_listener_impl.cc:114] TcpListener accepted 1 new connections.
[2024-12-20 17:29:30.048][76][trace][filter] [external/envoy/source/common/network/listener_filter_buffer_impl.cc:95] onFileEvent: 1
[2024-12-20 17:29:30.048][76][trace][filter] [external/envoy/source/common/network/listener_filter_buffer_impl.cc:60] recv returned: 0
[2024-12-20 17:29:30.048][76][trace][filter] [external/envoy/source/common/network/listener_filter_buffer_impl.cc:64] recv return try again
[2024-12-20 17:29:30.048][76][trace][filter] [external/envoy/source/common/network/listener_filter_buffer_impl.cc:95] onFileEvent: 1
[2024-12-20 17:29:30.048][76][trace][filter] [external/envoy/source/common/network/listener_filter_buffer_impl.cc:60] recv returned: 91
[2024-12-20 17:29:30.048][76][trace][filter] [external/envoy/source/extensions/filters/listener/tls_inspector/tls_inspector.cc:146] tls inspector: recv: 91
[2024-12-20 17:29:30.048][76][debug][misc] [cilium/bpf_metadata.cc:305] EGRESS POD IP: 10.20.254.178, destination IP: 10.10.0.33 sni: ""
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/conntrack.cc:178] cilium.bpf_metadata: Using conntrack map global
[2024-12-20 17:29:30.048][76][trace][filter] [cilium/conntrack.cc:197] cilium.bpf_metadata: Looking up key: a0a0021, a14feb2, 50, c797, 6, 0
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/conntrack.cc:128] cilium.bpf_metadata: Opened IPv4 conntrack map global
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/conntrack.cc:229] cilium.bpf_metadata: IPv4 conntrack map global lookup failed: No such file or directory
[2024-12-20 17:29:30.048][76][trace][filter] [cilium/ipcache.cc:102] cilium.ipcache: Looking up key: a14feb2, prefixlen: 32
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/ipcache.cc:110] cilium.ipcache: 10.20.254.178 has ID 2
[2024-12-20 17:29:30.048][76][trace][filter] [cilium/ipcache.cc:102] cilium.ipcache: Looking up key: a0a0021, prefixlen: 32
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/ipcache.cc:110] cilium.ipcache: 10.10.0.33 has ID 2
[2024-12-20 17:29:30.048][76][trace][filter] [cilium/ipcache.cc:102] cilium.ipcache: Looking up key: a2a0080, prefixlen: 32
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/ipcache.cc:110] cilium.ipcache: 10.42.0.128 has ID 8
[2024-12-20 17:29:30.048][76][debug][filter] [./cilium/socket_option.h:246] Cilium SocketOption(): source_identity: 8, ingress: false, port: 80, pod_ip: 10.42.0.128, source_addresses: /10.42.0.128:0/, mark: 80b00 (magic mark: b00, cluster: 0, ID: 8), proxy_id: 11889, sni: ""
[2024-12-20 17:29:30.048][76][trace][misc] [external/envoy/source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x1667fe95e000 for 3600000ms, min is 3600000ms
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/network_filter.cc:77] cilium.network: onNewConnection
[2024-12-20 17:29:30.048][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:474] [Tags: "ConnectionId":"574"] raising connection event 2
[2024-12-20 17:29:30.048][76][debug][conn_handler] [external/envoy/source/common/listener_manager/active_tcp_listener.cc:160] [Tags: "ConnectionId":"574"] new connection from 10.20.254.178:51095
[2024-12-20 17:29:30.048][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:228] item added to deferred deletion list (size=1)
[2024-12-20 17:29:30.048][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:122] clearing deferred deletion list (size=1)
[2024-12-20 17:29:30.048][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:619] [Tags: "ConnectionId":"574"] socket event: 3
[2024-12-20 17:29:30.048][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:742] [Tags: "ConnectionId":"574"] write ready
[2024-12-20 17:29:30.048][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:659] [Tags: "ConnectionId":"574"] read ready. dispatch_buffered_data=0
[2024-12-20 17:29:30.048][76][trace][connection] [external/envoy/source/common/network/raw_buffer_socket.cc:25] [Tags: "ConnectionId":"574"] read returns: 91
[2024-12-20 17:29:30.048][76][trace][connection] [external/envoy/source/common/network/raw_buffer_socket.cc:39] [Tags: "ConnectionId":"574"] read error: Resource temporarily unavailable, code: 0
[2024-12-20 17:29:30.048][76][trace][filter] [cilium/network_filter.cc:202] cilium.network: onData 91 bytes, end_stream: false
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:646] [Tags: "ConnectionId":"574"] parsing 91 bytes
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:588] [Tags: "ConnectionId":"574"] message begin
[2024-12-20 17:29:30.048][76][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:398] [Tags: "ConnectionId":"574"] new stream
[2024-12-20 17:29:30.048][76][trace][misc] [external/envoy/source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x1667fea24540 for 300000ms, min is 300000ms
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:545] [Tags: "ConnectionId":"574"] completed header: key=Host value=hello-world.onprem.internal
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:545] [Tags: "ConnectionId":"574"] completed header: key=User-Agent value=curl/8.7.1
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:839] [Tags: "ConnectionId":"574"] onHeadersCompleteImpl
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:545] [Tags: "ConnectionId":"574"] completed header: key=Accept value=*/*
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:1192] [Tags: "ConnectionId":"574"] Server: onHeadersComplete size=3
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:943] [Tags: "ConnectionId":"574"] message complete
[2024-12-20 17:29:30.048][76][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1147] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] request headers complete (end_stream=true):
':authority', 'hello-world.onprem.internal'
':path', '/'
':method', 'HEAD'
'user-agent', 'curl/8.7.1'
'accept', '*/*'
[2024-12-20 17:29:30.048][76][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1130] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] request end stream
[2024-12-20 17:29:30.048][76][debug][connection] [external/envoy/source/common/network/connection_impl.h:98] [Tags: "ConnectionId":"574"] current connecting state: false
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:603] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] decode headers called: filter=envoy.filters.http.grpc_web status=0
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:603] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] decode headers called: filter=envoy.filters.http.grpc_stats status=0
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/l7policy.cc:135] cilium.l7policy: downstream decodeHeaders()
[2024-12-20 17:29:30.048][76][trace][filter] [cilium/ipcache.cc:102] cilium.ipcache: Looking up key: a0a0021, prefixlen: 32
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/ipcache.cc:110] cilium.ipcache: 10.10.0.33 has ID 2
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/l7policy.cc:200] cilium.l7policy: Ingress from 2 policy lookup for endpoint 10.42.0.128 for port 80: ALLOW
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:603] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] decode headers called: filter=cilium.l7policy status=0
[2024-12-20 17:29:30.048][76][debug][router] [external/envoy/source/common/router/router.cc:447] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] no route match for URL '/'
[2024-12-20 17:29:30.048][76][debug][http] [external/envoy/source/common/http/filter_manager.cc:1027] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] Preparing local reply with details route_not_found
[2024-12-20 17:29:30.048][76][debug][http] [external/envoy/source/common/http/filter_manager.cc:1069] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] Executing sending local reply.
[2024-12-20 17:29:30.048][76][trace][misc] [external/envoy/source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x1667fea24540 for 300000ms, min is 300000ms
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:907] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] commonEncodePrefix end_stream: true, isHalfCloseEnabled: false
[2024-12-20 17:29:30.048][76][debug][filter] [cilium/l7policy.cc:246] cilium.l7policy: downstream encodeHeaders()
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:1245] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] encode headers called: filter=cilium.l7policy status=0
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:1245] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] encode headers called: filter=envoy.filters.http.grpc_stats status=0
[2024-12-20 17:29:30.048][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:1245] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] encode headers called: filter=envoy.filters.http.grpc_web status=0
[2024-12-20 17:29:30.048][76][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1838] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] encoding headers via codec (end_stream=true):
':status', '404'
'date', 'Fri, 20 Dec 2024 17:29:29 GMT'
'server', 'envoy'
[2024-12-20 17:29:30.049][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:534] [Tags: "ConnectionId":"574"] writing 106 bytes, end_stream false
[2024-12-20 17:29:30.049][76][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1950] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] Codec completed encoding stream.
[2024-12-20 17:29:30.049][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:228] item added to deferred deletion list (size=1)
[2024-12-20 17:29:30.049][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:228] item added to deferred deletion list (size=2)
[2024-12-20 17:29:30.049][76][trace][misc] [external/envoy/source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x1667fe95e000 for 3600000ms, min is 3600000ms
[2024-12-20 17:29:30.049][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:594] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] decodeHeaders filter iteration aborted due to local reply: filter=envoy.filters.http.router
[2024-12-20 17:29:30.049][76][trace][http] [external/envoy/source/common/http/filter_manager.cc:603] [Tags: "ConnectionId":"574","StreamId":"17226087540785556413"] decode headers called: filter=envoy.filters.http.router status=1
[2024-12-20 17:29:30.049][76][trace][http] [external/envoy/source/common/http/http1/codec_impl.cc:696] [Tags: "ConnectionId":"574"] parsed 91 bytes
[2024-12-20 17:29:30.049][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:122] clearing deferred deletion list (size=2)
[2024-12-20 17:29:30.049][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:619] [Tags: "ConnectionId":"574"] socket event: 2
[2024-12-20 17:29:30.049][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:742] [Tags: "ConnectionId":"574"] write ready
[2024-12-20 17:29:30.049][76][trace][connection] [external/envoy/source/common/network/raw_buffer_socket.cc:70] [Tags: "ConnectionId":"574"] write returns: 106
[2024-12-20 17:29:30.061][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:619] [Tags: "ConnectionId":"574"] socket event: 3
[2024-12-20 17:29:30.061][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:742] [Tags: "ConnectionId":"574"] write ready
[2024-12-20 17:29:30.061][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:659] [Tags: "ConnectionId":"574"] read ready. dispatch_buffered_data=0
[2024-12-20 17:29:30.061][76][trace][connection] [external/envoy/source/common/network/raw_buffer_socket.cc:25] [Tags: "ConnectionId":"574"] read returns: 0
[2024-12-20 17:29:30.061][76][debug][connection] [external/envoy/source/common/network/connection_impl.cc:719] [Tags: "ConnectionId":"574"] remote close
[2024-12-20 17:29:30.061][76][debug][connection] [external/envoy/source/common/network/connection_impl.cc:281] [Tags: "ConnectionId":"574"] closing socket: 0
[2024-12-20 17:29:30.075][76][trace][connection] [external/envoy/source/common/network/connection_impl.cc:474] [Tags: "ConnectionId":"574"] raising connection event 0
[2024-12-20 17:29:30.075][76][trace][conn_handler] [external/envoy/source/common/listener_manager/active_stream_listener_base.cc:126] [Tags: "ConnectionId":"574"] tcp connection on event 0
[2024-12-20 17:29:30.075][76][debug][conn_handler] [external/envoy/source/common/listener_manager/active_stream_listener_base.cc:136] [Tags: "ConnectionId":"574"] adding to cleanup list
[2024-12-20 17:29:30.075][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:228] item added to deferred deletion list (size=1)
[2024-12-20 17:29:30.075][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:228] item added to deferred deletion list (size=2)
[2024-12-20 17:29:30.075][76][trace][main] [external/envoy/source/common/event/dispatcher_impl.cc:122] clearing deferred deletion list (size=2)
```

### Anything else?

If I remove HTTPS listener section from gateway, I can access HTTP service again.

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by AurimasNav** [comment]
staying alive..

**Comment by DimitriosLisenko** [comment]
I am also seeing this issue - having trouble making cert-manager work with cilium's Gateway API implementation. Also seeing https://github.com/cilium/cilium/issues/34982 which also prevents cert-manager from working. Assuming they might be somehow related.

**Comment by mikejoh** [comment]
> I am also seeing this issue - having trouble making cert-manager work with cilium's Gateway API implementation. Also seeing [#34982](https://github.com/cilium/cilium/issues/34982) which also prevents cert-manager from working. I'm guessing that these two issues are related.
> 
> Removing the HTTPS listener + HTTPRoute and then adding them back in reverse order sometimes makes it so that the HTTP listener continues working, but then the challenge URL responds with a 404 instead - I'm guessing the internal state of the Envoy is broken at that point.

We're hitting exactly the same issue, it took a while for us to understand why the HTTPRoute never worked correctly. After removing the TLS parts altogether things started to work.

We're on:
* Cilium v1.17.0
* cert-manager v1.17.2
* GAPI v1.2

**Comment by youngnick** [comment]
Hi @AurimasNav, in the config you've given, you're right that the HTTP listener should keep working. Note that that HTTPRoute will never attach to the HTTPS listener though, because you've provided a SectionName. So it should only allow routing of HTTP traffic. With this config, the HTTPS listener will not be used. I suspect that's where this issue is arising - I don't think our testing covers this case.

**Comment by AurimasNav** [comment]
> Hi [@AurimasNav](https://github.com/AurimasNav), in the config you've given, you're right that the HTTP listener should keep working. Note that that HTTPRoute will never attach to the HTTPS listener though, because you've provided a SectionName. So it should only allow routing of HTTP traffic. With this config, the HTTPS listener will not be used. I suspect that's where this issue is arising - I don't think our testing covers this case.

Hi, thanks for the response, I'm aware that HTTPRoute will not attach to the HTTPS listener, this example is simply to demonstrate that HTTP listener stops working the moment I add HTTPS listener, in actuality I will have a separate HTTPRoute for HTTPS listener, but that was not necessary to reproduce this issue.

To elaborate on the actual use case - this bug prevents ACME HTTP01 challenge from completing successfully, because I can't have HTTP and HTTPS working simultaneously.

**Comment by jackmaninov** [comment]
> Hi, thanks for the response, I'm aware that HTTPRoute will not attach to the HTTPS listener, this example is simply to demonstrate that HTTP listener stops working the moment I add HTTPS listener, in actuality I will have a separate HTTPRoute for HTTPS listener, but that was not necessary to reproduce this issue.
> 
> To elaborate on the actual use case - this bug prevents ACME HTTP01 challenge from completing successfully, because I can't have HTTP and HTTPS working simultaneously.

In https://github.com/cilium/cilium/issues/34982#issuecomment-2972783391 a workaround for this issue has been found: separating the HTTP and HTTPS listeners across separate gateways. You'll need to use sharing keys if you would like both to attach to the same LB or IP.

**Comment by youngnick** [comment]
To be clear, I agree that there does seem to be a bug in Cilium somewhere here, but as I said on #34982, we may need to do some work on the testing before we can get to the bottom of it. Please use the workaround for now, and I'll come back once we see if we can fix this one.

**Comment by RarogCmex** [comment]
I hit this not only with Gateway API, but with Ingress API too. I can't add any new details, but it seems that Ingress is affected too.
Sadly I am unsure how to split http and https in my LAN at this moment, so I think whether I should adopt different Ingress/GW api provider. I'm subrscribed for the updates in this thread.

**Comment by youngnick** [comment]
I'd ask everyone running into this to please ensure you're testing with the most recent patch for Cilium v1.18, as we recently refactored the Gateway API reconciler in v1.18 and later, and that seems to have resolved the issues in #34982.

If you are seeing this issue, retesting using v1.18 and seeing if it still applies would be very helpful, thanks.

**Comment by RarogCmex** [comment]
@youngnick I have hit this issue with 1.18.4.

**Comment by pschiffe** [comment]
I'm still not able to issue Let's Encrypt cert with 1.18.4

**Comment by badnetmask** [comment]
I just upgraded to 1.18.5 and it seems to have worked for me.

**Comment by badnetmask** [comment]
> I just upgraded to 1.18.5 and it seems to have worked for me.

Sorry, I retract what I said. I wanted to test if this would work on a clean install, so I removed everything, and it did not work. The problem is still present even with 1.18.5 (at least for me).

**Comment by youngnick** [comment]
Okay, thanks everyone, it seems clear this issue is not resolved, we will look into it.

**Comment by repairman29** [comment]
🎯 **PING RECEIVED!**

**@repairman29** has claimed this bounty via [Echeo](https://echeo.io).

**Match Score:** 85%  
**Status:** Ready to start working!

[View conversation in Echeo](https://echeo.io/messages) to coordinate.

**Comment by repairman29** [comment]
Hey there! This looks like something I could help with. I'd be happy to take a look and see what we can do.

*(Found via [Echeo](https://echeo.io))*

**Comment by jackmaninov** [comment]
So I've long been affected by this issue as well (and currently working around it by using a separate Gateway to hold my HTTP listener). After puzzling for months and having claude-code dig into the code, I think the issue is a design gap in the hostname isolation logic when using Gateway API. This is based on my testing with v1.18.6.

Currently checkHostNameIsolation() correctly identifies that a more specific, secure route exists in cases like this, e.g. putting up an insecure listener for an ACME solver, or in my case I want to put up a redirect to HTTPS on the insecure port. The issue is that there is associated logic in operator/pkg/model/ingestion/gateway.go that prevents an *implicit* insecure listener attachment to a catch-all insecure listener, even when a route *explicitly* targets an insecure listener (i.e. with sectionName). 

Additionally, the Gateway's status REPORTS attachment of routes to the insecure listener, but the generated CEC does not include them (i.e. the relevant `'@type': type.gogleapis.com/envoy.config.route.v3.RouteConfiguration` for `listener-insecure` object will be blank unless there is some other insecure-only attachment). So this reported attachment is clearly erroneous; it should either not report it or there should be some separate .status report indicating an attachment was dropped due to host name isolation.

The reason why I say this appears to be a design gap is that you can currently "work around" this for Ingress resources with either the global `--enforce-ingress-https` flag or per-ingress `ingress.cilium.io/force-https` annotations; Cilium will build redirects to HTTPS for you. In my case I'm trying to manually build these HTTPS redirects with additional HTTPRoutes, but I need to use a separate Gateway with only the HTTP listener to be able to target them (if I try targeting by `sectionName` on the main Gateway they get dropped). In fact the "work-around" seems to be exactly what you would expect to happen, Cilium is dropping an insecure listener connection because it knows of a secure alternative, and for Ingresses it will generate a redirect to HTTPS for you; the gap is that this doesn't seem possible with Gateway API as well.

But in the case of Cert-Manager, it's trying to do an ACME challenge (and connect an insecure HTTPRoute to the acme solver pod it spins up) in response to an HTTPS listener that has *already* been added to the Gateway, but is either invalid (no TLS secret yet) or possibly improper (ACME needs to connect to you on HTTP). @youngnick recently refactored the GatewayAPI reconcilers in order to correct handling the invalid listener config case, but it doesn't fill the design gap. So in that case the insecure listener would have to be allowed to attach to the insecure listener, whether by allowing targeting the insecure listener explicitly (by section name, or HTTP protocol or port, etc) despite the secure listener that is also configured for the same hostname.

Hope that all makes sense, I'm probably missing additional logical context somewhere. This worked as expected back in Cilium 1.15, it looks like listener isolation (commit 7ab2a587d7 in PR #32233) is what made the change.

I'm running into this issue again because I now want to set my Gateways' `externalTrafficPolicy` to local, and this is incompatible with a split Gateway setup (as the LoadBalancer Services created for Gateways don't target envoy by backendRefs but instead by some magic Endpoints with 192.192.192.192 as address; I would need something like what issue https://github.com/cilium/cilium/issues/38766 proposes to proceed with this).

**Comment by youngnick** [comment]
Thanks for all of this detail @jackmaninov, that's excellent.

I've had a quick read, and I think I see what you're saying, although I'll need to look at the code more to confirm.

That said, I should also mention that `externalTrafficPolicy: Local` doesn't have the effect you might expect - I wrote a section in the docs about this a while ago: https://docs.cilium.io/en/stable/network/servicemesh/gateway-api/gateway-api/#externaltrafficpolicy-for-loadbalancer-or-nodeport-services

The key part is this:
> However, for Cilium’s ingress config, both Ingress and Gateway API, **it is not necessary** to configure `externalTrafficPolicy: Local` to keep the source IP visible to the backend pod (via the `X-Forwarded-For` and `X-Envoy-External-Address` fields).

**Comment by youngnick** [comment]
Reading over this again, I think a new test case is in order, which probably should end up as a conformance test upstream (since cert-manager is a common use case for Gateway API anyway). I will pick this up and see what I can do. No guarantees on timing though, sorry.

**Comment by jackmaninov** [comment]
Thanks for taking a look at this again, hopefully this is the core of the issue (and several others including https://github.com/cilium/cilium/issues/34982).

> That said, I should also mention that `externalTrafficPolicy: Local` doesn't have the effect you might expect - I wrote a section in the docs about this a while ago: https://docs.cilium.io/en/stable/network/servicemesh/gateway-api/gateway-api/#externaltrafficpolicy-for-loadbalancer-or-nodeport-services
> 
> The key part is this:
> 
> > However, for Cilium’s ingress config, both Ingress and Gateway API, **it is not necessary** to configure `externalTrafficPolicy: Local` to keep the source IP visible to the backend pod (via the `X-Forwarded-For` and `X-Envoy-External-Address` fields).

Yeah, my reason for doing this isn't source IP visibility but latency. I run geographically dispersed edge-nodes for my Gateways and I'm trying to avoid an extra hop between edge nodes in the LB->Envoy path. I could instead run Envoy with hostNetworking, but then I lose LB-IPAM.

**Comment by youngnick** [comment]
Hmm, Envoy in Daemonset mode always runs in host networking mode (the container, that is), so the traffic is _always_ local. It's not like another ingress controller, where the traffic _might_ get forwarded to another node. Every node has all the config, so traffic that arrives is always processed by Envoy on the node it arrived on. TPROXY can't do non-local destinations.

**Comment by youngnick** [comment]
I just opened #44492, which in my testing, will fix this issue. I'll arrange for that to be backported to v1.19 and v1.18 as well.

**Comment by youngnick** [comment]
#44492 has merged, so if anyone watching this could verify that the issue is now fixed in `main`, and update us here, then we can close this one out.

**Comment by jackmaninov** [comment]
I’m stuck on v1.18.6 until the kernel fix for #44216 lands in a Talos release (probably the next few days). After that I’ll take a look.

**Comment by tiborauer** [comment]
@youngnick, I believe it will be included in the next release (1.19.2?). Do you have an ETA?

**Comment by youngnick** [comment]
The backports to both v1.18 and v1.19 are completed, so this will be included in the next patch releases for each. I don't have an ETA at the moment though, sorry.

**Comment by joestringer** [comment]
Stable release cadence is documented here: https://docs.cilium.io/en/stable/contributing/release/organization/#stable-releases

**Comment by youngnick** [comment]
I'll leave this open until those releases are out and folks confirm that the issue is resolved.

**Comment by lopcode** [comment]
Managed to get certificates issued with a similar setup using `v1.19.2` (after experiencing this issue with `v1.19.1`) 🤞

**Comment by joestringer** [comment]
Thanks for the update! Closing fixed.
